### PR TITLE
Return migrated node template

### PR DESCRIPTION
**Pull request by rmweir**
**Problem:**
Node template migration breaks API for any client referring to legacy node templates.

**Solution:**
Forward all request for legacy node templates to their corresponding migrated node template.

**Issue:**
https://github.com/rancher/rancher/issues/26160

### CODE REVIEW

File: pkg/api/store/nodetemplate/store.go [hunk 1]
```diff
@@ -24,43 +24,95 @@ type nodeTemplateStore struct {
...
+	return fmt.Sprintf("%s:nt-%s-%s", namespace.NodeTemplateGlobalNamespace, ns, id)
+}
+
+func getNsAndID(id string) (string, string) {
+	idParts := strings.SplitN(id, ":", 2)
+	if len(idParts) == 1 {
+		return "", idParts[0]
+	}
+	return idParts[0], idParts[1]
+}
```
Comment (prachidamle) [inline pkg/api/store/nodetemplate/store.go:87]: we can use ref.Parse() utility method https://github.com/rancher/rancher/blob/master/pkg/ref/parse.go#L32

File: pkg/api/store/nodetemplate/store.go [hunk 2]
```diff
@@ -24,43 +24,95 @@ type nodeTemplateStore struct {
...
 	return data, nil
 }
 
+func (s *nodeTemplateStore) ByID(apiContext *types.APIContext, schema *types.Schema, id string) (map[string]interface{}, error) {
+	originalNs, originalID := getNsAndID(id)
+	fullMigratedID := getMigratedID(originalNs, originalID)
+	migratedNs, migratedID := getNsAndID(fullMigratedID)
+	data, err := s.Store.ByID(apiContext, schema, fullMigratedID)
+	if err != nil {
+		return nil, replaceIDInError(err, migratedID, originalID, migratedNs, originalNs)
```
Comment (prachidamle) [inline pkg/api/store/nodetemplate/store.go:65]: if s.Store.ByID returns say k8serror IsNotFound, their can be details about the legacy ID in the error, do we handle them too in replaceIDInError?

File: pkg/api/store/nodetemplate/store.go [hunk 3]
```diff
@@ -24,43 +24,95 @@ type nodeTemplateStore struct {
...
-		Store:       s,
-		Transformer: s.filterLegacyTemplates,
 	}
 }
 
-func (s *nodeTemplateStore) filterLegacyTemplates(apiContext *types.APIContext, schema *types.Schema, data map[string]interface{}, opt *types.QueryOptions) (map[string]interface{}, error) {
+func (t *transformer) filterLegacyTemplates(apiContext *types.APIContext, schema *types.Schema, data map[string]interface{}, opt *types.QueryOptions) (map[string]interface{}, error) {
 	ns, _ := data["namespaceId"].(string)
-	name, _ := data["name"].(string)
+	id, _ := data["id"].(string)
```
Comment (prachidamle) [inline pkg/api/store/nodetemplate/store.go:50]: can you pls explain why is this change needed?

File: pkg/api/store/nodetemplate/store.go [hunk 4]
```diff
@@ -24,43 +24,95 @@ type nodeTemplateStore struct {
...
+		msg = strings.Replace(msg, ns, newNs, -1)
+		err = httperror.NewAPIError(apiError.Code, msg)
+	}
+	return err
+}
+
 func (s *nodeTemplateStore) Delete(apiContext *types.APIContext, schema *types.Schema, id string) (map[string]interface{}, error) {
+	originalNs, originalID := getNsAndID(id)
+	fullMigratedID := getMigratedID(originalNs, originalID)
+	migratedNs, migratedID := getNsAndID(fullMigratedID)
```
Comment (prachidamle) [inline pkg/api/store/nodetemplate/store.go:101]: Can we move the three steps to a separate method and call that instead in all the Store API's?

File: pkg/api/store/nodetemplate/store.go [hunk 5]
```diff
@@ -24,43 +24,95 @@ type nodeTemplateStore struct {
...
+		if pool.Spec.NodeTemplateName == fullMigratedID {
 			return nil, httperror.NewAPIError(httperror.MethodNotAllowed, "Template is in use by a node pool.")
 		}
 	}
-	return s.Store.Delete(apiContext, schema, id)
+	data, err := s.Store.Delete(apiContext, schema, fullMigratedID)
+	if err != nil {
+		return nil, replaceIDInError(err, migratedID, originalID, migratedNs, originalNs)
+	}
+	return data, nil
```
Comment (prachidamle) [inline pkg/api/store/nodetemplate/store.go:115]: replace back the originalId and NS in the response data?

File: pkg/api/store/nodetemplate/store.go (hunk 3 above)
Comment (rmweir) [reply pkg/api/store/nodetemplate/store.go:50]: name was the display name, so nothing which is not valid for enqueueing

File: pkg/api/store/nodetemplate/store.go (hunk 5 above)
Comment (rmweir) [reply pkg/api/store/nodetemplate/store.go:115]: yes, id is being replaced to not break anything that may rely or reuse response id- as request by @ibuildthecloud and @cloudnautique

Comment (prachidamle) [reply pkg/api/store/nodetemplate/store.go:115]: right understood, but I meant why are we not replacing it back here for Delete() like its done in Update()

Comment (rmweir) [reply pkg/api/store/nodetemplate/store.go:115]: oh- because data return by deletes is nil anyway. I can just return a nil to be safe

File: pkg/api/store/nodetemplate/store.go (hunk 2 above)
Comment (rmweir) [reply pkg/api/store/nodetemplate/store.go:65]: I believe in the proxy store k8s errors (which fit the interface Status) are translated to norman api errors: https://github.com/rancher/norman/blob/07a7713a06d92f6a2aee25cb2b86c6670577febc/store/proxy/error_wrapper.go#L46

File: pkg/api/store/nodetemplate/store.go [hunk 6]
```diff
@@ -24,43 +24,98 @@ type nodeTemplateStore struct {
...
+	}
+
+	data["namespaceId"] = originalNs
+	data["id"] = id
+	return data, nil
+}
+
+// getAllIDs returns the namespace and trimmed ID of the original ID
+// and its corresponding migrated ID
+func getAllIDs(id string) (string, string, string, string, string) {
```
Comment (dramich) [inline pkg/api/store/nodetemplate/store.go:73]: I think this could just return a struct with the fields, returning 5 strings from one func is pretty rough to reason about.

File: tests/integration/suite/test_node_templates.py [hunk 7]
```diff
@@ -47,16 +52,57 @@ def migrated_template_exists(id):
...
+    def legacy_template_deleted():
+        try:
+            k8s_dynamic_client.get_namespaced_custom_object(
+                "management.cattle.io", "v3", ns.name, 'nodetemplates',
+                legacy_id)
+            return False
+        except ApiException as e:
+            assert e.status == 404
+            print(e)
+            return True
```
Comment (dramich) [inline tests/integration/suite/test_node_templates.py:76]: In place of the True, do a check on the error code, if it's 404 return true, if not let it keep going and time out. Also drop the print

File: tests/integration/suite/test_node_templates.py [hunk 8]
```diff
@@ -47,16 +52,57 @@ def migrated_template_exists(id):
...
+
+    # wait for legacy node template to be deleted
+    wait_for(lambda: legacy_template_deleted(), fail_handler=lambda:
              "failed waiting for old node template to delete")
 
+    # retrieve node template via legacy id
+    nt = admin_client.by_id_node_template(id=full_legacy_id)
```
Comment (dramich) [inline tests/integration/suite/test_node_templates.py:83]: Validate using both IDs get the same data back before attempting to change it.

File: tests/integration/suite/test_node_templates.py [hunk 9]
```diff
@@ -47,16 +52,57 @@ def migrated_template_exists(id):
...
+    nt2 = admin_client.by_id_node_template(id=id)
+    # assert node template being updated is migrated node template
+    assert nt2.azureConfig.customData == new_config.customData
+
+    # delete node template via legacy id
+    admin_client.delete(nt)
+    wait_for(lambda: admin_client.by_id_node_template(id) is None,
+             fail_handler=lambda:
+             "failed waiting for migrate node template to delete")
+    admin_client.by_id_node_template(id="something:soment")
```
Comment (dramich) [inline tests/integration/suite/test_node_templates.py:104]: This looks unfinished, is there going to be another check?

**Comment by rmweir** [comment]
Tested using 1.8.1 of rancher2 provider. When upgrading from 2.3.2 to master and using `terraform apply`, the node template is recreated because it thinks it does not exist. When upgrading from 2.3.2 to this PR and using `terraform apply`, nothing changes
